---
title: "parallelnewhybrid"
output:
  html_document:
    theme: yeti
    highlight: tango
---


### ***parallelnewhybrid*** is an R package designed to parallelize *NewHybrids* analyses



###Overview of functions:  

**Function name** | **Main use**  
------------|---------------------------------------------------------------------------
**parallelnh_xx.R** | Iteratively assigns *NewHybrids* analyses to the *c* available cores in a computer.  
**nh_zcore.R** | Inserts genotype category assignments to *NewHybrids* format files (see Anderson 2002)  
**freqbasedsim.R** | Creates simulated hybrid datasets comprised of *two pure parental groups*, *F1*, *F2*, and *backcrosses to both parentals* 
**freqbasedsim_Advanced.R** | Creates simulated hybrid datasets comprised of *two pure parental groups*, *F1*, *F2*, and *backcrosses to both ancestral groups*. Allows the number of individuals of each genotype frequency category output to be specified by the user. 
**freqbasedsim_UB2.R** | Creates simulated hybrid datasets comprised of *two pure parental groups*, *F1*, *F2*, and *backcrosses to both ancestral groups*. A less biased version of the simulation function which allows the user to specify the proportion of individuals in each of the two ancestral populations to be used to generate the allele frequencies upon which to base the simulations. Also allows the number of individuals of each genotype frequency category output to be specified by the user. 

***  

###Package Installation
**parallelenh** can be installed from GitHub using the R package *devtools* and the function *install_github*:  
```
devtools::install_github("bwringe/parallelnewhybrid")
```
<span style = "color:red"> <strong>NOTE:</strong></span> : **parallelnh** relies on functions from the R packages *parallel*, *plyr*, *stringr*, and *tidyr*. The user should ensure these are installed from CRAN prior to installing **parallelnh**.  

***  

###Function descriptions
<h4 class="text-primary">parallelnh_xx.R</h4>
Allows *NewHybrids* (Anderson and Thompson 2002) to be run in parallel. A job (*NewHybrids* analysis) is assigned to each of the *c* cores available in the computer. As each task finishes, a new analysis is asigned to the idled core.  
All *NewHybrids* format files in the folder the user specifies as *folder.data* will be analyzed.  
The user can also specify the length of the MCMC burnin using the *burnin* and *sweeps* parameters.  
<span style = "color:red"> <strong>NOTE:</strong></span> : There are **three operating system-specific versions** of the **parallelnh_xx** function because of the different ways in which the operating systems handle forking of processes.  

**parallelnh version**|**Operating system**
------------|----------
**parallelnh_OSX** | OS X
**parallelnh_WIN** | Windows
**parallelnh_LINUX** | Linux (Ubuntu)

<h4 class="text-primary">nh_zcore.R</h4>
The programme *NewHybrids* allows for individuals to be designated as belonging to a known genotype category to help build the allele frequencies for the various genotype classes. The function *nh_zcore* will insert a column of known genotype categories (which we call Zvec) into a *NewHybrids* formatted dataset, then output a *NewHybrids* format file, with a name indicating that the category vector has been added. Additionally, whether the known individuals should be included in the analysis, or used only to generate the genotype frequencies of the various gentype categories can also be specified by adding an 's' to the Zvec (i.e. z0s).   
A single Zvec can be applied to all files in a folder in a batch fashion, or each file can be given a unique Zvec.  
For information on the known genotype categories, refer to the *NewHybrids* documentation (Anderson 2003). The numbering in the Zvec should follow described in the *NewHybrids* documentation (e.g. Pure1 = z0, Pure2 = z1, F1 = z2 etc.).  

<h4 class="text-primary">freqbasedsim.R</h4>
This function generates simulated, centred Pure1, Pure2, F1, F2, BC1 and BC2 offspring based on the genotype frequencies of two ancestral populations provided in a *GENEPOP* [link](http://genepop.curtin.edu.au/) formatted file. The user can specify how many simulated files to create, how many individuals to simulate, and how many replicate files to create.  

***  

###How to use:  
####Example datasets:  
Example datasets have been provided as R images (.rda files). These can be loaded into your workspace using the ``data()`` function  

**Example dataset** | **Contents**  
------------|---------------------------------------------------------------  
*PurePops* | A *GENEPOP* format file containing the genotypes at 240 SNP loci for 60 individuals in each of two ancestral populations (PopA and PopB). This is the data format for **freqbasedsim**, and as such the file can be entered into this function directly.  
*SimPops\_S1R1_NH* | A *NewHybrids* format file. This file is the result of running **freqbasedsim** using the default parameters on the *PurePops* dataset. *SimPops\_S1R1_NH* can be analyzed using the function **parallelnh_xx**.  
*Zed\_vec* | A two column dataset of Zeds to be inserted into the dataset *SimPops\_S1R1_NH* using the function **nh_zcore**  

***  
<h4 class="text-primary">parallelnh_xx</h4>  

**Parameter** | **Description**
------------|---------------------------------------------------------------  
*folder.data*| A file path to the folder in which the *NewHybrids* formatted files to be analyzed reside.   
*where.NH* | A file path to the *NewHybrids* installation folder
*burnin* | An integer specifying how many burnin steps *NewHybrids* is to run
*sweeps* | An integer specifying how many sweep steps *NewHybrids* is to run  

```r
### Run analyses on all NewHybrids format files in the folder "your_data_to_analyze"
parallelnh_OSX(folder.data = "~/ ... /your_data_to_analyze/", where.NH = "~/ ... /newhybrids/", burnin = 5000, sweeps = 10000)
## the usage would be the same for the WIN and LINUX versions.
```

***  

### **parallelnewhybrid** Contributors

* Brendan Wringe <https://github.com/bwringe>; bwringe(at)gmail.com; Corresponding author  
* Dr. Ian Bradbury <http://tinyurl.com/h3qrhkv>  
* Dr. Nick Jeffery <https://github.com/NickJeff13>  
* Dr. Ryan Ronald Edward Stanley <https://github.com/rystanley> 

* If you don't understand anything, please contact me (bwringe[at]gmail.com)
* Any ideas on how to improve the functionality are very much appreciated.
* If you spot a typo, feel free to edit and send a pull request.  

To cite the current version of **parallelnewhybrid**, please refer to our zenodo DOI: 

***
***