---
title: "parallelnewhybrid"
output:
  html_document:
    theme: yeti
    highlight: tango
---


### ***parallelnewhybrid*** is an R package designed to parallelize *NewHybrids* analyses, and can also rapidly create multi-generation simulated hybrid datasets
***



###Overview of functions:  

**Function name** | **Main use**  
------------|---------------------------------------------------------------------------
**parallelnh_xx.R** | Iteratively assigns *NewHybrids* analyses to the *c* available cores in a computer.  
**nh_zcore.R** | Inserts genotype category assignments to *NewHybrids* format files (see Anderson 2002)  
**freqbasedsim.R** | Creates simulated hybrid datasets comprised of *two pure parental groups*, *F1*, *F2*, and *backcrosses to both parentals* 
**freqbasedsim_Advanced.R** | Creates simulated hybrid datasets comprised of *two pure parental groups*, *F1*, *F2*, and *backcrosses to both ancestral groups*. Allows the number of individuals of each genotype frequency category output to be specified by the user. 
**freqbasedsim_UB2.R** | Creates simulated hybrid datasets comprised of *two pure parental groups*, *F1*, *F2*, and *backcrosses to both ancestral groups*. A less biased version of the simulation function which allows the user to specify the proportion of individuals in each of the two ancestral populations to be used to generate the allele frequencies upon which to base the simulations. Also allows the number of individuals of each genotype frequency category output to be specified by the user. 

***  

###Package Installation
**parallelenh** can be installed from GitHub using the R package *devtools* and the function *install_github*:  
```
devtools::install_github("bwringe/parallelnewhybrid")
```
<span style = "color:red"> <strong>NOTE:</strong></span> : **parallelnh** relies on functions from the R packages *parallel*, *plyr*, *stringr*, and *tidyr*. The user should ensure these are installed from CRAN prior to installing **parallelnh**.  

***  

###Function descriptions
<h4 class="text-primary">parallelnh_xx.R</h4>
Allows *NewHybrids* (Anderson and Thompson 2002) to be run in parallel. A job (*NewHybrids* analysis) is assigned to each of the *c* cores available in the computer. As each task finishes, a new analysis is asigned to the idled core.  
All *NewHybrids* format files in the folder the user specifies as *folder.data* will be analyzed.  
The user can also specify the length of the MCMC burnin using the *burnin* and *sweeps* parameters.  
<span style = "color:red"> <strong>NOTE:</strong></span> : There are **three operating system-specific versions** of the **parallelnh_xx** function because of the different ways in which the operating systems handle forking of processes.  

**parallelnh version**|**Operating system**
------------|----------
**parallelnh_OSX** | OS X
**parallelnh_WIN** | Windows
**parallelnh_LINUX** | Linux (Ubuntu)

<h4 class="text-primary">nh_zcore.R</h4>
The programme *NewHybrids* allows for individuals to be designated as belonging to a known genotype category to help build the allele frequencies for the various genotype classes. The function *nh_zcore* will insert a column of known genotype categories (which we call Zvec) into a *NewHybrids* formatted dataset, then output a *NewHybrids* format file, with a name indicating that the category vector has been added. Additionally, whether the known individuals should be included in the analysis, or used only to generate the genotype frequencies of the various gentype categories can also be specified by adding an 's' to the Zvec (i.e. z0s).   
A single Zvec can be applied to all files in a folder in a batch fashion, or each file can be given a unique Zvec.  
For information on the known genotype categories, refer to the *NewHybrids* documentation (Anderson 2003). The numbering in the Zvec should follow described in the *NewHybrids* documentation (e.g. Pure1 = z0, Pure2 = z1, F1 = z2 etc.).  

<h4 class="text-primary">freqbasedsim.R</h4>
This function generates simulated, centred Pure1, Pure2, F1, F2, BC1 and BC2 offspring based on the genotype frequencies of two ancestral populations provided in a *GENEPOP* [link](http://genepop.curtin.edu.au/) formatted file. The user can specify how many simulated files to create, how many individuals to simulate, and how many replicate files to create.  

***  

###How to use:  
####Example datasets:  
Example datasets have been provided as R images (.rda files). These can be loaded into your workspace using the ``data()`` function  

**Example dataset** | **Contents**  
------------|---------------------------------------------------------------  
*PurePops* | A *GENEPOP* format file containing the genotypes at 240 SNP loci for 60 individuals in each of two ancestral populations (PopA and PopB). This is the data format for **freqbasedsim**, and as such the file can be entered into this function directly.  
*SimPops\_S1R1_NH* | A *NewHybrids* format file. This file is the result of running **freqbasedsim** using the default parameters on the *PurePops* dataset. *SimPops\_S1R1_NH* can be analyzed using the function **parallelnh_xx**.  
*Zed\_vec* | A two column dataset of Zeds to be inserted into the dataset *SimPops\_S1R1_NH* using the function **nh_zcore**  

***  
<h4 class="text-primary">parallelnh_xx</h4>  

**Parameter** | **Description**
------------|---------------------------------------------------------------  
*folder.data*| A file path to the folder in which the *NewHybrids* formatted files to be analyzed reside.   
*where.NH* | A file path to the *NewHybrids* installation folder
*burnin* | An integer specifying how many burnin steps *NewHybrids* is to run
*sweeps* | An integer specifying how many sweep steps *NewHybrids* is to run  

```r
### Run analyses on all NewHybrids format files in the folder "your_data_to_analyze"
parallelnh_OSX(folder.data = "~/ ... /your_data_to_analyze/", where.NH = "~/ ... /newhybrids/", burnin = 5000, sweeps = 10000)
## the usage would be the same for the WIN and LINUX versions.
```

***  

<h4 class="text-primary">freqbasedsim</h4>  

**Parameter** | **Description**
------------|---------------------------------------------------------------  
*GenePopData*| GenePop formatted file containing genotypes from two (2) ancestral populations. This is the data from which the simulated hybrids will be constructed  
*pop.groups*|An optional character vector denoting how the two ancestral populations should be named; default is PopA and PopB
*outputName*|An optional character vector to be applied as the name of the output. The default is NULL, in which case the output name is constructed from the name of the input, with the suffix *\_SiRj_NH* added where *i* is the number of *S*imulations corresponding to the output, and *j* is the number of *R*eplicates of the *ith* simulation. NH refers to the fact that the output is in *NewHybrids* format.  
*sample.size*|An optional integer to specify the number of simulated individuals to be created for each of the six hybrid classes (e.g. 200 * each of Pure1, Pure2, F1, F2, BC1 and BC2 = 1200 total simulated individuals). Default is 200  
*NumSims*|An integer number of simulated datasets to be created. Default is 1  
*NumReps*|An integer number of replicates to be created for each of the *i* simulated datasets specified in NumSims. Default is 1  

```r
### Create a simulated dataset comprised of 200 individuals in each of the six genotype frequency classes (i.e. Pure1, Pure2, F1, F2, BC1 and BC2)

freqbasedsim(GenePopData = "~/ ... /YourReferenceData.txt", sample.size = 200, NumSims = 3, NumReps = 3)
```

***  

<h4 class="text-primary">freqbasedsim_Advanced</h4>  

**Parameter** | **Description**
------------|---------------------------------------------------------------  
*GenePopData*| GenePop formatted file containing genotypes from two (2) ancestral populations. This is the data from which the simulated hybrids will be constructed  
*pop.groups*|An optional character vector denoting how the two ancestral populations should be named; default is PopA and PopB
*outputName*|An optioanal character vector to be applied as the name of the output. The default is NULL, in which case the output name is constructed from the name of the input, with the suffix *\_SiRj_NH* added where *i* is the number of *S*imulations corresponding to the output, and *j* is the number of *R*eplicates of the *ith* simulation. NH refers to the fact that the output is in *NewHybrids* format.  
*sample.sizePure*|An optional integer to specify the number of simulated  Pure1 and Pure2 individuals (centred ancestral populations) to be created. Default is 200  
*sample.sizeF1*|An optional integer to specify the number of simulated  F1 individuals (Pure1 X Pure2) to be created. Default is 200  
*sample.sizeF2*|An optional integer to specify the number of simulated  F2 dividuals (F1 X F1) to be created. Default is 200  
*sample.sizeBC*|An optional integer to specify the number of simulated  Back Cross1 and BackCross2 individuals (Pure1 X F1 and Pure2XF2) to be created. Default is 200  
*NumSims*|An integer number of simulated datasets to be created. Default is 1  
*NumReps*|An integer number of replicates to be created for each of the *i* simulated datasets specified in NumSims. Default is 1  

```r
## If the number of Pure, F1, F2 and Backcross individuals to be returned is not specified, this function is the same as freqbased sim. Because NewHybrids utilizes a Bayesian algorithm, results may be influenced by the proportions of individuals of different genotypic frequency categories provided, thus this function can be used to tailor the numbers of simulated individuals in each category to those expected of your study system.

### Suppose you expect that the majority of individuals in will be pure, with the numbers of Pure > F1 > F2 > BC

freqbasedsim_Advanced(GenePopData = , sample.sizePure = 100, sample.sizeF1 = 50, sample.sizeF2 = 10, sample.sizeBC = 10, NumSims = 3, NumReps = 3)

## This will create three independent simulations, each replicated three times.
```
***

<h4 class="text-primary">freqbasedsim_UB2</h4>

**Parameter** | **Description**
------------|---------------------------------------------------------------  
*GenePopData* | GenePop formatted file containing genotypes from two (2) ancestral populations. This is the data from which the simulated hybrids will be constructed  
*pop.groups* | An optional character vector denoting how the two ancestral populations should be named. Default is PopA and PopB  
*prop.sample* | The proportion of individuals in both ancestral PopA and PopB to sample to create the simulated hybrids (Pure1, Pure2, F1, F2, BC1, and BC2)  
*outputName* | An optioanal character vector to be applied as the name of the output. The default is NULL, in which case the output name is constructed from the name of the input, with the suffix *\_SiRj_NH* added where *i* is the number of *S*imulations corresponding to the output, and *j* is the number of *R*eplicates of the *ith* simulation. NH refers to the fact that the output is in *NewHybrids* format.  
*sample.sizePure1* | An optional integer to specify the number of simulated Pure1 individuals (centred ancestral PopA) to be output when the desired number is less than the number of individuals in Ancestral Population 1 * prop.sample. The default is NULL, where the number output = number of individuals in Ancestral Population 1 * prop.sample. If a number greater than number of individuals in Ancestral Population 1 * prop.sample is requested, the number of individuals in Ancestral Population 1 * prop.sample are output.  
*sample.sizePure2* | An optional integer to specify the number of simulated Pure2 individuals (centred ancestral PopB) to be output when the desired number is less than the number of individuals in Ancestral Population 2 * prop.sample. The default is NULL, where the number output = number of individuals in Ancestral Population 2 * prop.sample. If a number greater than number of individuals in Ancestral Population 2 * prop.sample is requested, the number of individuals in Ancestral Population 2 * prop.sample are output.  
*sample.sizeF1* | An optional integer to specify the number of F1 individuals to be output when the desired number of simulated F1 individuals is less than (number of individuals in Ancestral PopA + number of individuals in Ancestral PopB) * prop.sample. The default is NULL where the number returned = (number of individuals in Ancestral PopA + number of individuals in Ancestral PopB) * prop.sample. Unless sample.sizeF1 is explicitly stated, even when sample.sizePure1 and sample.sizePure2 are specified, the number of simulated F1 individuals returned will be equal to (number of individuals in Ancestral PopA + number of individuals in Ancestral PopB) * prop.sample.  
*sample.sizeF2* | An optional integer to specify the number of simulated F2 individuals to be output when the desired number of simulated F2 individuals is less than (number of individuals in Ancestral PopA + number of individuals in Ancestral PopB) * prop.sample. The default is NULL where the number returned = (number of individuals in Ancestral PopA + number of individuals in Ancestral PopB) * prop.sample. Unless sample.sizeF2 is explicitly stated, even when sample.sizePure1 and sample.sizePure2 are specified, the number of simulated F2 individuals returned will be equal to (number of individuals in Ancestral PopA + number of individuals in Ancestral PopB) * prop.sample.  
*sample.sizeBC1* | An optional integer to specify the number of simulated BC1 (PopA X F1) individuals to be output when the desired number of simulated BC1 individuals is less than the number of individuals in Ancestral PopA * prop.sample. The default is NULL where the number returned = number of individuals in Ancestral PopA * prop.sample. Unless sample.sizeBC1 is explicitly stated, even when sample.sizePure1 and sample.sizeF1 are specified, the number of simulated BC1 individuals returned will be equal to number of individuals in Ancestral PopA * prop.sample.  
*sample.sizeBC2* | An optional integer to specify the number of simulated BC2 (PopB X F1) individuals to be output when the desired number of simulated BC2 individuals is less than the number of individuals in Ancestral PopB * prop.sample. The default is NULL where the number returned = number of individuals in Ancestral PopB * prop.sample. Unless sample.sizeBC2 is explicitly stated, even when sample.sizePure2 and sample.sizeF1 are specified, the number of simulated BC2 individuals returned will be equal to number of individuals in Ancestral PopB * prop.sample.  
*NumSims* | An integer number of simulated datasets to be created. Default is 1  
*NumReps* | An integer number of replicates to be created for each of the *i* simulated datasets specified in NumSims. Default is 1  

```r
## This function is designed to be more unbiased than the other simulation functions. This is accomplished by sampling, without replacement, only a proportion of the individuals provided (Anderson 2010). 

freqbasedsim_UB2(GPD = "~/ ... /YourReferenceData.txt")
```

***

<h4 class="text-primary">nh_zcore</h4>  


**Parameter** | **Description**
------------|---------------------------------------------------------------  
*GetstheZdir*| A file path to the folder in which the *NewHybrids* formatted files to be appended reside
*multiapplyZvec*| **IF** *a single Zvec is to be applied to each **NewHybrids** format file*, use *multiapplyZvec*. Zvec will added in batch fashion to every *NewHybrids* format file within a working folder. *multiapplyZvec* must be specified as a file path + file name (e.g. "~/HoldenUniversity/CoolEthansComputer/HairDolls/Zvexxx.csv")
*applyuniqueZvec*| **IF** *each **NewHybrids** format file is to be given a UNIQUE vector of Zvecs* use *applyuniqueZvec*. <span style = "color:red"> <strong>NOTE:</strong></span> : to apply UNIQUE Zvecs, the Zvec files must all be placed in a single folder separate from the *NewHybrids* files *AND* they MUST follow the file name convention "NHFileName_Zvec.csv" where NHFileName is the same name as the file to which the Zvec is to be applied. Consequently, the number of *NewHybrids* files = number of Zvec files. *applyuniqueZvec* must be specified as a file path to the folder in which the Zvec files reside (e.g. "~/HarrisonUniversity/DEANGordonPritchard/LettersOfRecommendation/")  

```r
## to apply the same Zed scores to multiple files.

nh_zcore(GetstheZdir = "~/ ... /yourNHFiles/", multiapplyZvec = "~/ ... /yourZeds.txt")  ## NOTE: Single file

## to apply different Zed scores to different files

nh_zcore(GetstheZdir = "~/ ... /yourNHFiles/", applyuniqueZvec = "~/ ... /yourZedFiles/") ## Zed files must all be placed in a single folder separate from the NewHybrids files AND they MUST follow the file name convention "NHFileName_Zvec.csv" where NHFileName is the same name as the file to which the Zvec is to be applied
```


